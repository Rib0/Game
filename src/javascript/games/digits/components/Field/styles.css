.field {
    width: 100%;
    max-width: 400px;
    height: 400px;
    margin: 30px auto;
    position: relative;
    border: 1px solid black;
}

.cell {
    display: inline-block;
    width: 25%;
    border: 1px solid black;
    height: 100px;
    position: absolute;
    top: 0;
    left: 0;
    opacity: 0;
    transition-property: top, left, opacity;
    transition-duration: 0.3s;
}
