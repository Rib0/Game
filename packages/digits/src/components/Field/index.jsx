import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import MainView from '@/components/MainView';
import styles from './styles.css';

export class Field extends PureComponent {
    static Cell({ children, style, onClick, data }) {
        return (
            <div className={styles.cell} onClick={onClick} style={style} {...data}>
                {children}
            </div>
        );
    }

    getCellStyles = index => {
        const { loading } = this.props;
        const sideSize = MainView.SideSize.length;
        const leftMulti = index % sideSize;
        const topMulti = parseInt(index / sideSize, 10);

        return {
            top: loading ? 0 : `${topMulti * MainView.CellSize}px`,
            left: loading ? 0 : `${leftMulti * MainView.CellSize}px`,
            opacity: index < sideSize ** 2 - 1 && loading ? 0 : 1,
        };
    };

    render() {
        const { field, onClick } = this.props;

        return (
            <div className={styles.field}>
                {field.map(({ value, order }) => {
                    const style = {
                        ...this.getCellStyles(order),
                        backgroundColor: value - order === 1 && 'lightgreen',
                    };

                    return (
                        <Field.Cell
                            style={style}
                            onClick={onClick}
                            data={{
                                'data-index': order,
                            }}
                            key={value}
                        >
                            <div className={styles.value}>{value}</div>
                        </Field.Cell>
                    );
                })}
            </div>
        );
    }
}

Field.propTypes = {
    field: PropTypes.arrayOf(PropTypes.shape({
        value: PropTypes.string,
        order: PropTypes.string,
    })),
    onClick: PropTypes.func,
    loading: PropTypes.bool,
};