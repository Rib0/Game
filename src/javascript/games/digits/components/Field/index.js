import React, { PureComponent } from 'react';

import styles from './styles.css';

class Field extends PureComponent {
    static Row({ children }) {
        return <div className={styles.row}>{children}</div>;
    }

    static Col({ children }) {
        return (
            <div className={styles.col} onClick={onClick}>
                {children}
            </div>
        );
    }

    render() {
        const { children } = this.props;

        return <div className={styles.field}>{children}</div>;
    }
}

export default Field;
