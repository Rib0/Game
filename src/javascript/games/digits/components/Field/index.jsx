import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import styles from './styles.css';

class Field extends PureComponent {
    static Cell({ children, style, onClick, data }) {
        return (
            <div className={styles.cell} onClick={onClick} style={style} {...data}>
                {children}
            </div>
        );
    }

    render() {
        const { children } = this.props;

        return <div className={styles.field}>{children}</div>;
    }
}

Field.propTypes = {
    children: PropTypes.element,
};

export default Field;
