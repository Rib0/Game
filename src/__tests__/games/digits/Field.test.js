import React from 'react';
import Field from 'games/digits/components/Field';

const setUp = props => shallow(<Field {...props} />);

describe('Digits game - Field test', () => {
    const props = {
        field: [
            { value: 2, order: 1 },
            { value: 6, order: 2 },
            { value: 3, order: 3 },
            ,
            { value: 4, order: 4 },
            { value: 5, order: 5 },
        ],
        loading: false,
        onClick: jest.fn(),
    };

    test('Field should render', () => {
        const Component = setUp(props);
        expect(Component).toMatchSnapshot();
    });
    test('Field should have 5 cells', () => {
        const Component = setUp(props);
        expect(Component.find('Cell')).toHaveLength(5);
    });
    test('Callback should call if field clicked', () => {
        const Component = setUp(props);
        const cell = Component.find('Cell').first();
        cell.simulate('click');
        expect(props.onClick).toHaveBeenCalledTimes(1);
    });
    test('Field static method Cell should render', () => {
        const Component = shallow(
            <Field.Cell>
                <div />
                <div />
            </Field.Cell>
        );
        const cells = Component.find('div');
        expect(cells).toHaveLength(3);
    });
    test('Cells should be in the corner if loading is true', () => {
        const newProps = {
            ...props,
            loading: true,
        };

        const Component = setUp(newProps).instance();
        const resultStyles = Component.getCellStyles(2);
        expect(resultStyles).toEqual({
            top: 0,
            left: 0,
            opacity: 0,
        });
    });
    test('Cell in right order should have lightgreen color', () => {
        const Component = setUp(props);
        const cell = Component.find('Cell').first();
        expect(cell.prop('style')).toEqual({
            ...Component.instance().getCellStyles(props.field[0].order),
            backgroundColor: 'lightgreen',
        });
    });
});
