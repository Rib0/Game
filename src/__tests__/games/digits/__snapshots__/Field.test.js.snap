// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Digits game - Field test Field should render 1`] = `
<div
  className="field"
>
  <Cell
    data={
      Object {
        "data-index": 1,
      }
    }
    key="2"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": "lightgreen",
        "left": "100px",
        "opacity": 1,
        "top": "0px",
      }
    }
  >
    <div
      className="value"
    >
      2
    </div>
  </Cell>
  <Cell
    data={
      Object {
        "data-index": 2,
      }
    }
    key="6"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": false,
        "left": "200px",
        "opacity": 1,
        "top": "0px",
      }
    }
  >
    <div
      className="value"
    >
      6
    </div>
  </Cell>
  <Cell
    data={
      Object {
        "data-index": 3,
      }
    }
    key="3"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": false,
        "left": "300px",
        "opacity": 1,
        "top": "0px",
      }
    }
  >
    <div
      className="value"
    >
      3
    </div>
  </Cell>
  <Cell
    data={
      Object {
        "data-index": 4,
      }
    }
    key="4"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": false,
        "left": "0px",
        "opacity": 1,
        "top": "100px",
      }
    }
  >
    <div
      className="value"
    >
      4
    </div>
  </Cell>
  <Cell
    data={
      Object {
        "data-index": 5,
      }
    }
    key="5"
    onClick={[MockFunction]}
    style={
      Object {
        "backgroundColor": false,
        "left": "100px",
        "opacity": 1,
        "top": "100px",
      }
    }
  >
    <div
      className="value"
    >
      5
    </div>
  </Cell>
</div>
`;
